[22704] Link problem with MinGW

### Is there an already existing issue for this?

- [x] I have searched the existing issues

### Expected behavior

I have already compile foonathan_memory, fastcdr and tinyxml2 with MinGW successfully. But there are some problem when I compile the fastdds with MinGW, like "i64 to LL","stat redefined","declspec( dllexport )  to __attribute__((visibility("default")))"...... I used some special means to solve these problems and finally successfully compiled fastdds, then i  got libfastdds.dll and libfastdds.dll.a


### Current behavior

But when I use libfastdds.dll and libfastdds.dll.a in my project, there are some thing wrong like
`E:/mingw32-14/bin/../lib/gcc/i686-w64-mingw32/14.2.0/../../../../i686-w64-mingw32/bin/ld.exe: CMakeFiles\MBus.dir/objects.a(startlogPubSubTypes.cxx.obj):startlogPubSubTypes.cxx:(.text+0x5cd): undefined reference to `eprosima::fastdds::MD5::MD5()'`......
All the symbol from FastDDS which i used in my project was undefined. I tried every import and export combination in fastdds_dll.hpp, "__attribute__((visibility("default")))" "declspec( dllexport )" "declspec( dllimport )" "null" . But does not work.
Are there some special export config for MinGW in FastDDS?

### Steps to reproduce

1:Compile foonathan_memory, fastcdr and tinyxml2 with MinGW.
2:Modify the source code, like modify CMAKE_CXX_FLAGS, modify include in some files, and so on.
3:Use libfastdds.dll.a in your project.

### Fast DDS version/commit

FastDDS v3.1.1 master
FastCDR v2.2.5 
MinGW  v14.2.0
Windows10

### Platform/Architecture

Other. Please specify in Additional context section.

### Transport layer

Default configuration, UDPv4 & SHM

### Additional context

_No response_

### XML configuration file

_No response_

### Relevant log output

_No response_

### Network traffic capture

_No response_

@Javgilavi 
Yes, my pleasure. “solution means” can be divided into three parts. 
First, warning with little impact. like "unused-parameter" "missing-field-initializers" and so on. I just use command like "-Wno-unused-parameter" to ignore them.
Then, some llibrary error between windows and unix.  Added option for "elif defined(MINGW)" to ifdef _WIN32(or _MSC_VER) and else , then fill the MINGW branch, so it can work as same as the _WIN32(or _MSC_VER)  branch. For example, in "RobustExclusiveLock.hpp", add  "elif defined(MINGW)" to ifdef _MSC_VER and else in the "include" block and "class" block. Then copy the code from "_MSC_VER" block, use the same function "_sopen_s" with <io.h>(MinGW), modify "_SH_DENYRW" to "0x0010", which is not defined in  <io.h>(MinGW).
Third, Manually link library, like "target_link_libraries(fastdds PUBLIC ws2_32 mswsock)".

As for the error in this issue, I solved this problem with "declspec( dllexport )" and "-Wno-attributes". So i can use almost all the symbols from libfastdds.dll and libfastdds.dll.a. 
Why "almost all"?  There are some new problem for "TypeSupport". 
When I use `TypeSupport(new startlogPubSubType())` to construct a TypeSupport object, i got the error
` E:/mingw32-14/bin/../lib/gcc/i686-w64-mingw32/14.2.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\Users\CUILIU~1\AppData\Local\Temp\ccEr7V0I.o:xmsgbus.cxx:(.text$_ZN8eprosima7fastdds3dds11TypeSupportC1EPNS1_13TopicDataTypeE[__ZN8eprosima7fastdds3dds11TypeSupportC1EPNS1_13TopicDataTypeE]+0x20): undefined reference to `vtable for eprosima::fastdds::dds::TypeSupport'     `
Maybe there are some thing wrong with "class TypeSupport : public std::shared_ptr<TopicDataType>" and MinGW?
By the way, "startlogPubSubType" was generated by fastdds-gen with a idl.

> [@Javgilavi](https://github.com/Javgilavi) Yes, my pleasure. “solution means” can be divided into three parts. First, warning with little impact. like "unused-parameter" "missing-field-initializers" and so on. I just use command like "-Wno-unused-parameter" to ignore them. Then, some llibrary error between windows and unix. Added option for "elif defined(MINGW)" to ifdef _WIN32(or _MSC_VER) and else , then fill the MINGW branch, so it can work as same as the _WIN32(or _MSC_VER) branch. For example, in "RobustExclusiveLock.hpp", add "elif defined(MINGW)" to ifdef _MSC_VER and else in the "include" block and "class" block. Then copy the code from "_MSC_VER" block, use the same function "_sopen_s" with <io.h>(MinGW), modify "_SH_DENYRW" to "0x0010", which is not defined in <io.h>(MinGW). Third, Manually link library, like "target_link_libraries(fastdds PUBLIC ws2_32 mswsock)".
> 
> As for the error in this issue, I solved this problem with "declspec( dllexport )" and "-Wno-attributes". So i can use almost all the symbols from libfastdds.dll and libfastdds.dll.a. Why "almost all"? There are some new problem for "TypeSupport". When I use `TypeSupport(new startlogPubSubType())` to construct a TypeSupport object, i got the error `E:/mingw32-14/bin/../lib/gcc/i686-w64-mingw32/14.2.0/../../../../i686-w64-mingw32/bin/ld.exe: C:\Users\CUILIU~1\AppData\Local\Temp\ccEr7V0I.o:xmsgbus.cxx:(.text$_ZN8eprosima7fastdds3dds11TypeSupportC1EPNS1_13TopicDataTypeE[__ZN8eprosima7fastdds3dds11TypeSupportC1EPNS1_13TopicDataTypeE]+0x20): undefined reference to "vtable for eprosima::fastdds::dds::TypeSupport" ` Maybe there are some thing wrong with "class TypeSupport : public std::shared_ptr<TopicDataType>" and MinGW? By the way, "startlogPubSubType" was generated by fastdds-gen with a idl.

I have hit the same walls as you hence reproducible and have reached the same point in compilation. Where do we go from here?

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -395,7 +395,26 @@ endif()
 ###############################################################################
 # Tools default setup
 ###############################################################################
-option(COMPILE_TOOLS "Build tools" ON)
+if(WIN32 AND CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+    execute_process(
+        COMMAND ${CMAKE_CXX_COMPILER} -dumpmachine
+        OUTPUT_VARIABLE COMPILER_MACHINE
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+    if(COMPILER_MACHINE MATCHES "mingw")
+        message(STATUS "Using MinGW compiler.")
+        option(COMPILE_TOOLS "Build tools" OFF)
+        add_definitions(-DMINGW_COMPILER=1)
+        set(CMAKE_CXX_FLAGS
+            "${CMAKE_CXX_FLAGS} -Wno-attributes -Wno-stringop-overread -Wno-builtin-macro-redefined -Wno-cast-function-type -Wno-unused-variable -Wno-unused-parameter -Wno-unused-function -Wno-missing-field-initializers")
+        target_link_libraries(fastdds PUBLIC ws2_32 mswsock)
+    else()
+        message(STATUS "Using a GNU compiler on Windows but not MinGW.")
+    endif()
+else()
+    message(STATUS "Not using a GNU compiler on Windows.")
+    option(COMPILE_TOOLS "Build tools" ON)
+endif()
 
 if(EPROSIMA_BUILD)
     set(COMPILE_TOOLS ON)

--- a/include/fastdds/fastdds_dll.hpp
+++ b/include/fastdds/fastdds_dll.hpp
@@ -44,11 +44,19 @@
 
 #if defined(_WIN32)
 #if defined(EPROSIMA_ALL_DYN_LINK) || defined(FASTDDS_DYN_LINK)
-#if defined(fastdds_EXPORTS)
-#define FASTDDS_EXPORTED_API __declspec( dllexport )
+#if defined(MINGW_COMPILER)
+    #if defined(fastdds_EXPORTS)
+    #define FASTDDS_EXPORTED_API __declspec( dllexport )
+    #else
+    #define FASTDDS_EXPORTED_API __attribute__((visibility("default")))
+    #endif // FASTDDS_SOURCE
 #else
-#define FASTDDS_EXPORTED_API __declspec( dllimport )
-#endif // FASTDDS_SOURCE
+    #if defined(fastdds_EXPORTS)
+    #define FASTDDS_EXPORTED_API __declspec( dllexport )
+    #else
+    #define FASTDDS_EXPORTED_API __declspec( dllimport )
+    #endif // FASTDDS_SOURCE
+#endif // if defined(MINGW_COMPILER)
 #else
 #define FASTDDS_EXPORTED_API
 #endif // if defined(EPROSIMA_ALL_DYN_LINK) || defined(FASTDDS_DYN_LINK)

--- a/include/fastdds/utils/TimedMutex.hpp
+++ b/include/fastdds/utils/TimedMutex.hpp
@@ -26,6 +26,8 @@
 
 #if defined(_MSC_FULL_VER) && _MSC_FULL_VER >= 193632528
 #include <mutex>
+#elif defined(MINGW_COMPILER)
+#include <mutex>
 #else
 #include <thread>
 extern int clock_gettime(
@@ -45,6 +47,9 @@ namespace fastdds {
 #if defined(_WIN32)
 
 #if defined(_MSC_FULL_VER) && _MSC_FULL_VER >= 193632528
+using TimedMutex = std::timed_mutex;
+using RecursiveTimedMutex = std::recursive_timed_mutex;
+#elif defined(MINGW_COMPILER)
 using TimedMutex = std::timed_mutex;
 using RecursiveTimedMutex = std::recursive_timed_mutex;
 #else

--- a/src/cpp/fastdds/xtypes/serializers/json/dynamic_data_json.cpp
+++ b/src/cpp/fastdds/xtypes/serializers/json/dynamic_data_json.cpp
@@ -792,7 +792,8 @@ ReturnCode_t json_serialize_basic_member(
 #ifdef MINGW_COMPILER
                 std::string utf8_value;
                 int size_needed = std::wcstombs(nullptr, value.data(), 0);
-                if (size_needed > 0) {
+                if (size_needed > 0) 
+                {
                     utf8_value.resize(size_needed);
                     std::wcstombs(&utf8_value[0], value.data(), size_needed);
                 }

--- a/src/cpp/fastdds/xtypes/serializers/json/dynamic_data_json.cpp
+++ b/src/cpp/fastdds/xtypes/serializers/json/dynamic_data_json.cpp
@@ -750,7 +750,8 @@ ReturnCode_t json_serialize_basic_member(
                 std::wstring aux_wstring_value({value});
                 std::string utf8_value;
                 int size_needed = std::wcstombs(nullptr, aux_wstring_value.data(), 0);
-                if (size_needed > 0) {
+                if (size_needed > 0)
+                {
                     utf8_value.resize(size_needed);
                     std::wcstombs(&utf8_value[0], aux_wstring_value.data(), size_needed);
                 }
@@ -759,7 +760,7 @@ ReturnCode_t json_serialize_basic_member(
                 std::wstring_convert<std::codecvt_utf8<wchar_t>> converter;
                 std::string utf8_value = converter.to_bytes(aux_wstring_value);
 
-#endif  // defined(MINGW_COMPILER)              
+#endif  // defined(MINGW_COMPILER)
                 json_insert(member_name, utf8_value, output);
             }
             else
@@ -792,7 +793,7 @@ ReturnCode_t json_serialize_basic_member(
 #ifdef MINGW_COMPILER
                 std::string utf8_value;
                 int size_needed = std::wcstombs(nullptr, value.data(), 0);
-                if (size_needed > 0) 
+                if (size_needed > 0)
                 {
                     utf8_value.resize(size_needed);
                     std::wcstombs(&utf8_value[0], value.data(), size_needed);

--- a/src/cpp/fastdds/xtypes/serializers/json/dynamic_data_json.cpp
+++ b/src/cpp/fastdds/xtypes/serializers/json/dynamic_data_json.cpp
@@ -13,7 +13,9 @@
 // limitations under the License.
 
 #include <bitset>
-#include <codecvt>
+#ifndef MINGW_COMPILER
+    #include <codecvt>
+#endif  // ifndef MINGW_COMPILER
 #include <iomanip>
 #include <iostream>
 #include <string>
@@ -744,9 +746,20 @@ ReturnCode_t json_serialize_basic_member(
             if (RETCODE_OK == ret)
             {
                 // Insert UTF-8 converted value
+#if defined(MINGW_COMPILER)
+                std::wstring aux_wstring_value({value});
+                std::string utf8_value;
+                int size_needed = std::wcstombs(nullptr, aux_wstring_value.data(), 0);
+                if (size_needed > 0) {
+                    utf8_value.resize(size_needed);
+                    std::wcstombs(&utf8_value[0], aux_wstring_value.data(), size_needed);
+                }
+#else
                 std::wstring aux_wstring_value({value});
                 std::wstring_convert<std::codecvt_utf8<wchar_t>> converter;
                 std::string utf8_value = converter.to_bytes(aux_wstring_value);
+
+#endif  // defined(MINGW_COMPILER)              
                 json_insert(member_name, utf8_value, output);
             }
             else
@@ -776,8 +789,17 @@ ReturnCode_t json_serialize_basic_member(
             if (RETCODE_OK == ret)
             {
                 // Insert UTF-8 converted value
+#ifdef MINGW_COMPILER
+                std::string utf8_value;
+                int size_needed = std::wcstombs(nullptr, value.data(), 0);
+                if (size_needed > 0) {
+                    utf8_value.resize(size_needed);
+                    std::wcstombs(&utf8_value[0], value.data(), size_needed);
+                }
+#else
                 std::wstring_convert<std::codecvt_utf8<wchar_t>> converter;
                 std::string utf8_value = converter.to_bytes(value);
+#endif  // defined(MINGW_COMPILER)
                 json_insert(member_name, utf8_value, output);
             }
             else

--- a/src/cpp/utils/SystemInfo.cpp
+++ b/src/cpp/utils/SystemInfo.cpp
@@ -155,9 +155,17 @@ fastdds::dds::ReturnCode_t SystemInfo::get_username(
 bool SystemInfo::file_exists(
         const std::string& filename)
 {
+#ifdef _WIN32
+    // modify for mingw
+    DWORD fileAttributes = GetFileAttributesA(filename.c_str());
+    if (fileAttributes == INVALID_FILE_ATTRIBUTES) {
+        return false;
+    }
+    return !(fileAttributes & FILE_ATTRIBUTE_DIRECTORY);
+#else
     struct stat s;
-    // Check existence and that it is a regular file (and not a folder)
     return (stat(filename.c_str(), &s) == 0 && s.st_mode & S_IFREG);
+#endif
 }
 
 bool SystemInfo::wait_for_file_closure(

--- a/src/cpp/utils/SystemInfo.cpp
+++ b/src/cpp/utils/SystemInfo.cpp
@@ -158,7 +158,7 @@ bool SystemInfo::file_exists(
 #ifdef _WIN32
     // modify for mingw
     DWORD fileAttributes = GetFileAttributesA(filename.c_str());
-    if (fileAttributes == INVALID_FILE_ATTRIBUTES) 
+    if (fileAttributes == INVALID_FILE_ATTRIBUTES)
     {
         return false;
     }

--- a/src/cpp/utils/SystemInfo.cpp
+++ b/src/cpp/utils/SystemInfo.cpp
@@ -158,14 +158,15 @@ bool SystemInfo::file_exists(
 #ifdef _WIN32
     // modify for mingw
     DWORD fileAttributes = GetFileAttributesA(filename.c_str());
-    if (fileAttributes == INVALID_FILE_ATTRIBUTES) {
+    if (fileAttributes == INVALID_FILE_ATTRIBUTES) 
+    {
         return false;
     }
     return !(fileAttributes & FILE_ATTRIBUTE_DIRECTORY);
 #else
     struct stat s;
     return (stat(filename.c_str(), &s) == 0 && s.st_mode & S_IFREG);
-#endif
+#endif // ifdef _WIN32
 }
 
 bool SystemInfo::wait_for_file_closure(

--- a/src/cpp/utils/TimedConditionVariable.cpp
+++ b/src/cpp/utils/TimedConditionVariable.cpp
@@ -50,9 +50,15 @@
     return 0;
    }
  */
-#define exp7           10000000i64     //1E+7     //C-file part
-#define exp9         1000000000i64     //1E+9
-#define w2ux 116444736000000000i64     //1.jan1601 to 1.jan1970
+#ifdef MINGW_COMPILER
+    #define exp7           10000000LL     //1E+7     //C-file part
+    #define exp9         1000000000LL     //1E+9
+    #define w2ux 116444736000000000LL     //1.jan1601 to 1.jan1970
+#else
+    #define exp7           10000000i64     //1E+7     //C-file part
+    #define exp9         1000000000i64     //1E+9
+    #define w2ux 116444736000000000i64     //1.jan1601 to 1.jan1970
+#endif
 void unix_time(
         struct timespec* spec)
 {

--- a/src/cpp/utils/TimedConditionVariable.cpp
+++ b/src/cpp/utils/TimedConditionVariable.cpp
@@ -58,7 +58,7 @@
     #define exp7           10000000i64     //1E+7     //C-file part
     #define exp9         1000000000i64     //1E+9
     #define w2ux 116444736000000000i64     //1.jan1601 to 1.jan1970
-#endif
+#endif // ifdef MINGW_COMPILER
 void unix_time(
         struct timespec* spec)
 {

--- a/src/cpp/utils/shared_memory/RobustExclusiveLock.hpp
+++ b/src/cpp/utils/shared_memory/RobustExclusiveLock.hpp
@@ -17,6 +17,8 @@
 
 #ifdef  _MSC_VER
 #include <io.h>
+#elif defined(MINGW_COMPILER)
+#include <io.h>
 #else
 #include <sys/file.h>
 #endif // ifdef  _MSC_VER
@@ -142,6 +144,45 @@ private:
 
         int fd;
         ret = _sopen_s(&fd, file_path.c_str(), O_CREAT | O_RDONLY, _SH_DENYRW, _S_IREAD | _S_IWRITE);
+
+        if (ret != 0)
+        {
+            return -1;
+        }
+
+        *was_lock_created = true;
+
+        return fd;
+    }
+
+    static void unlock_and_close(
+            int fd,
+            const std::string& name)
+    {
+        _close(fd);
+
+        if (0 != std::remove(SharedDir::get_lock_path(name).c_str()))
+        {
+            EPROSIMA_LOG_WARNING(RTPS_TRANSPORT_SHM, "Failed to remove " << SharedDir::get_lock_path(name));
+        }
+    }
+
+#elif defined(MINGW_COMPILER)
+    static int open_and_lock_file(
+            const std::string& file_path,
+            bool* was_lock_created)
+    {
+        int test_exist;
+        auto ret = _sopen_s(&test_exist, file_path.c_str(), O_RDONLY, 0x0010, _S_IREAD | _S_IWRITE);
+
+        if (ret == 0)
+        {
+            *was_lock_created = false;
+            return test_exist;
+        }
+
+        int fd;
+        ret = _sopen_s(&fd, file_path.c_str(), O_CREAT | O_RDONLY, 0x0010, _S_IREAD | _S_IWRITE);
 
         if (ret != 0)
         {

--- a/src/cpp/utils/shared_memory/RobustSharedLock.hpp
+++ b/src/cpp/utils/shared_memory/RobustSharedLock.hpp
@@ -17,6 +17,8 @@
 
 #ifdef  _MSC_VER
 #include <io.h>
+#elif defined(MINGW_COMPILER)
+#include <io.h>
 #else
 #include <sys/file.h>
 #endif // ifdef  _MSC_VER
@@ -205,6 +207,118 @@ private:
 
             // Lock exclusive
             ret = _sopen_s(&fd, file_path.c_str(), _O_WRONLY, _SH_DENYWR, _S_IREAD | _S_IWRITE);
+            if (ret != 0)
+            {
+                lock_status = LockStatus::LOCKED;
+            }
+            else
+            {
+                _close(fd);
+            }
+        }
+        else
+        {
+            lock_status = LockStatus::OPEN_FAILED;
+        }
+
+        if (lock_status == LockStatus::NOT_LOCKED && remove_if_unlocked)
+        {
+            if (0 != std::remove(file_path.c_str()))
+            {
+                EPROSIMA_LOG_WARNING(RTPS_TRANSPORT_SHM, "Failed to remove " << file_path);
+            }
+        }
+
+        return lock_status;
+    }
+
+#elif defined(MINGW_COMPILER)
+    int open_and_lock_file(
+            const std::string& file_path,
+            bool* was_lock_created,
+            bool* was_lock_released)
+    {
+        int test_exist;
+
+        // Try open exclusive
+        auto ret = _sopen_s(&test_exist, file_path.c_str(), _O_WRONLY, 0x0010, _S_IREAD | _S_IWRITE);
+
+        if (ret == 0)
+        {
+            *was_lock_created = false;
+
+            if (was_lock_released)
+            {
+                *was_lock_released = true;
+            }
+
+            _close(test_exist);
+        }
+        else
+        {
+            // Try open shared
+            ret = _sopen_s(&test_exist, file_path.c_str(), _O_RDONLY, 0x0010, _S_IREAD | _S_IWRITE);
+
+            if (ret == 0)
+            {
+                if (was_lock_released)
+                {
+                    *was_lock_released = false;
+                }
+
+                *was_lock_created = false;
+
+                return test_exist;
+            }
+            else
+            {
+                if (was_lock_released)
+                {
+                    *was_lock_released = true;
+                }
+
+                *was_lock_created = true;
+            }
+        }
+
+        int fd;
+        // Open or create shared
+        ret = _sopen_s(&fd, file_path.c_str(), O_CREAT | _O_RDONLY, 0x0010, _S_IREAD | _S_IWRITE);
+
+        if (ret != 0)
+        {
+            char errmsg[1024];
+            strerror_s(errmsg, sizeof(errmsg), errno);
+            throw std::runtime_error("failed to open/create " + file_path + " " + std::string(errmsg));
+        }
+
+        return fd;
+    }
+
+    void unlock_and_close()
+    {
+        _close(fd_);
+
+        test_lock(SharedDir::get_lock_path(name_), true);
+    }
+
+    static LockStatus test_lock(
+            const std::string& file_path,
+            bool remove_if_unlocked = false)
+    {
+        LockStatus lock_status;
+
+        int fd;
+        auto ret = _sopen_s(&fd, file_path.c_str(), _O_RDONLY, 0x0010, _S_IREAD | _S_IWRITE);
+
+        if (ret == 0)
+        {
+            lock_status = LockStatus::NOT_LOCKED;
+
+            _close(fd);
+
+            // Lock exclusive
+            ret = _sopen_s(&fd, file_path.c_str(), _O_WRONLY, 0x0010, _S_IREAD | _S_IWRITE);
             if (ret != 0)
             {
                 lock_status = LockStatus::LOCKED;

--- a/thirdparty/filewatch/FileWatch.hpp
+++ b/thirdparty/filewatch/FileWatch.hpp
@@ -30,7 +30,9 @@
 
 #ifdef _WIN32
 #define WIN32_LEAN_AND_MEAN
-#define stat _stat
+#ifndef MINGW_COMPILER
+    #define stat _stat  // do not use stat in windows
+#endif  // ifndef MINGW_COMPILER
 #ifndef NOMINMAX
 #define NOMINMAX
 #endif
@@ -404,7 +406,7 @@ namespace filewatch {
                                 std::ratio_multiply<std::hecto, typename std::chrono::nanoseconds::period>>(
                                     reinterpret_cast<ULARGE_INTEGER*>(&att.ftLastWriteTime)->QuadPart - base_.first.QuadPart);
 
-                    if (bytes_returned == 0 || (current_time == last_write_time_) && current_size == last_size_ ) {
+                    if (bytes_returned == 0 || ((current_time == last_write_time_) && current_size == last_size_ )) {
                         break;
                     }
 

